TPP-00067: Cycle Service AsyncStorage Endpoints

kenneth-miura opened the pull request:
# Description

* Implements API as specified in the cycle section of the [wiki](https://github.com/uoftblueprint/the-period-purse/wiki/Onboarding,-Cycle,-Info,-Settings-API) in the file `CycleService.js`.
* Installed `date-fns` library for date comparison & difference between days.
* Updated Wiki to explicitly document expected behaviour for periods that span 2 different years

Fixes #67 

## Type of change

Please delete options that are not relevant.


- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (if required)
- [x] My changes generate no new warnings

kenneth-miura committed:
simple gets + post

kenneth-miura committed:
intermediary commit

kenneth-miura committed:
getPeriodDay done, getMostRecent testing

kenneth-miura committed:
added docs

kenneth-miura committed:
integrating new Symptoms class

kenneth-miura committed:
Merge branch 'develop' into tpp-00067

kenneth-miura committed:
move utils + swap to using utils in CycleService

kenneth-miura committed:
working getPeriodDay and getMostRecentPeriod

kenneth-miura committed:
added GetCycleDonutPercent

kenneth-miura committed:
implemented history, having issues with date localization

kenneth-miura committed:
new getsymptoms + day diff

kenneth-miura committed:
Merge branch 'develop' into tpp-00067

kenneth-miura committed:
cycle history now doesn't cut off for cross-year

kenneth-miura committed:
first and last interval now carries over years

kenneth-miura committed:
removed testing page + debugging prints

kenneth-miura committed:
stopped defaulting to removed test page on startup

faye-tan commented on tpp-app/src/services/utils/helpers.js:
```diff
@@ -16,3 +18,29 @@ const initializeEmptyYear = (yearNumber) => {
+/**
+ * Convert a Date object into a date string, encoding year, month and day. Note it encodes months as 1 indexed, and days as 0 indexed
+ * @param {Date} date Object to convert to string
+ * @return {String} String encoding year, month and day
```
YYYY-MM-DD format? I would add to comments

faye-tan commented on tpp-app/src/services/utils/helpers.js:
```diff
@@ -16,3 +18,29 @@ const initializeEmptyYear = (yearNumber) => {
+/**
+ * Retrieves the user's symptom data for the given date.
+ * @param day number
+ * @param month number (January = 0)
```
I think for arguments, we would pass 1 for January

kenneth-miura replied:
oh yeah u right

faye-tan commented on tpp-app/src/services/cycle/CycleService.js:
```diff
@@ -0,0 +1,279 @@
+ * @param {Date} laterDate
+ * @return {number} number of days between the two dates provided, ignoring their hours, minutes and seconds.
+ */
+const getDaysDiff = (earlierDate, laterDate) => {
```
i would maybe put this in utils since it does involve AsyncStorage and other places may need as well

kenneth-miura replied:
I wasn't sure if I should do that since it's not really computing the difference in days. So if you put in "Jan 21" and "Jan 22" it returns 2. 

I was worried if I put it in helpers people would use it assuming it returns the numeric difference between the date (so "Jan 21" and "Jan 22" returns 1) w/o really reading the docstring. I think a better name would help but I'm not sure what name conveys what the function is doing

annie951977 replied:
Maybe like getDaysDiffInclusive?

faye-tan commented on tpp-app/src/services/utils/helpers.js:
```diff
@@ -16,3 +18,29 @@ const initializeEmptyYear = (yearNumber) => {
+  const yearData = JSON.parse(await AsyncStorage.getItem(year.toString()));
+
+  // Return symptoms for that day or empty symptoms object if it doesn't exist
+  return yearData[month-1][day-1] ? new Symptoms(yearData[month-1][day-1]) : new Symptoms();
```
Loving this! so if something goes wrong, return Symptoms obj with everything as null. No unexpected surprises. 

One Q: if the passed in day month and year are invalid what will happen? Just in case we should guard against an IndexOutOfBounds when accessing ```yearData[month-1][day-1]```

kenneth-miura replied:
I'll talk it over with Emily but I would think returning `Symptoms` with everything null works? I think that we can usually treat the UNRECORDED state and the INVALID state for a day's symptoms the same way

faye-tan commented on tpp-app/src/services/cycle/CycleService.js:
```diff
@@ -0,0 +1,279 @@
+    try {
+      // Search backwards until date switches to the previous year
+      while(current.getFullYear() === year){
+        let currentSymptoms = await GETsymptomsForDate(current.getDate(), current.getMonth()+1, current.getFullYear());
```
I think we should load an entire year's worth of data from asyncStorage once and then traverse it rather than traverse while calling GETSymptomsForDate potentially ~365 times. So calling GETSymptomsForDate() gets the below structure which you can traverse through: 

```
[
    [
         Symptom { 
             “Flow”:  “LIGHT”,
             “Mood”: “HAPPY”,
             "Sleep": "7.5", 
             "Cramps": "MEDIUM", 
             "Exercise: null,
             "Notes": null
         }, 
         Symptom { 
             “Flow”:  null
             “Mood”: null,
             "Sleep": null, 
             "Cramps": null, 
             "Exercise: null,
             "Notes": null
         }, 
         ..., 
    ], 
    [...], 
    ...
]
```

Propose we add a helper method ```GETSymptomsForYear(year: number)``` to helpers.js where we use .map() to turn every day entry in the Calendar into a Symptoms obj. Then traverse through it. Map docs: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map 
```
 // This is not tested so syntax may not be 100% but the general structure is here
export const GETSymptomsForYear = async (year) => {
  // get year
  const yearData = JSON.parse(await AsyncStorage.getItem(year.toString()));

  // Check since yearData could be null 
  if (yearData) {
     // .map() basically transforms each item in the array. In this case we turn it all to Symptom objs
      yearData.map((month) => {
          month.map((day) => {
              return day ? new Symptoms(day) : new Symptoms();
          })
      })
     return yearData; 
  }
  return null;
}
```

This would be a pretty big upheaval from what you have now, open to discuss further! 

@edwardhuahan you probably have to traverse through calendar as well to display filter views, do you think this helper function would help/do you have something similar

kenneth-miura replied:
oh yeah that's def better. I'll make the swap

kenneth-miura committed:
Revert "removed testing page + debugging prints"

This reverts commit d5a233f0755c19e31d667c70bab6d2b4035e8aa8.

kenneth-miura committed:
switched to using getCalendarByYear

kenneth-miura committed:
swapped over to using get symptoms by year

kenneth-miura committed:
renamed functions + deleted testing stuff

kenneth-miura committed:
removed GETSymptomsByDate

helen-li commented on tpp-app/src/services/cycle/CycleService.js:
```diff
@@ -0,0 +1,304 @@
+import AsyncStorage from '@react-native-async-storage/async-storage';
+import {FLOW_LEVEL} from '../utils/constants';
+import {initializeEmptyYear, getDateString, GETsymptomsFor getCalendarByYear, getSymptomsFromCalendar}  from '../utils/helpers';
```
prob a typo here with GETsymptomsFor?

helen-li commented on tpp-app/src/services/cycle/CycleService.js:
```diff
@@ -0,0 +1,304 @@
+      return 0;
+    }
+    else {
+      let startDate = await this.GETMostRecentPeriodStartDay();
```
just curious if there was a reason to call `GETMostRecentPeriodStartDay` here rather than calling `getLastPeriodStart` directly with today's date via `new Date()`?

kenneth-miura replied:
I used `GETMostRecentPeriodStartDay` b/c it's in the defined API so I already had it. It's functionally equivalent to calling `getLastPeriodStart` directly with today's date via new Date() like u said, but shorter. Do you think we need both of the functions, since there's a pretty significant overlap in functionality?

helen-li replied:
oo i see hmmm personally it feels a bit redundant to me to have both but i also get it if we just keep it as is

kenneth-miura committed:
updated to add default calendars

kenneth-miura committed:
switched from rejects

helen-li commented on tpp-app/src/services/cycle/CycleService.js:
```diff
@@ -0,0 +1,309 @@
+        }
+
+        var yesterday = new Date(current.getTime())
+        yesterday.setDate(current.getDate() - 1)
```
something minor but these two lines are used quite often throughout - may be beneficial to modularize them into a simple helper to return the date object we need? beyond taking in the current date as a param the helper can also take in a number (refers to how many days we want before or after the current date) so that we can also use the helper beyond just for getting yesterday but also for getting dates like twoDaysLater, tomorrow, twoDaysEarlier, etc as needed in the methods at the top of the file

kenneth-miura committed:
moved days diff to helper

kenneth-miura committed:
refactored to use date-fns day diff fns

kenneth-miura committed:
Merge branch 'develop' into tpp-00067

kenneth-miura committed:
remove extra export on Symptoms

annie951977 commented on tpp-app/src/services/cycle/CycleService.js:
```diff
@@ -0,0 +1,290 @@
+   * Get how far the user is into their period as a percentage
+   * @return {Promise} A percentage approximation (meaning in range [0,1]) of how far the user is into their period
+   */
+  GETCycleDonutPercent: async function() {
```
I love that it's still the cycle donut. This is it. This is my legacy

annie951977 commented on tpp-app/src/services/cycle/CycleService.js:
```diff
@@ -0,0 +1,290 @@
+      noFlowTomorrow = (tomorrowSymptoms.flow === FLOW_LEVEL.NONE || tomorrowSymptoms.flow === null) ;
+      flowTwoDaysLater = (twoDaysLaterSymptoms.flow !== null && twoDaysLaterSymptoms.flow !== FLOW_LEVEL.NONE)
+    }
+
```
Love the way you're iterating through the dates

annie951977 reviewed (approved):
LGTM! Good job

faye-tan commented on tpp-app/src/services/cycle/CycleService.js:
```diff
@@ -0,0 +1,290 @@
+
+
+
+    //return twoDaysEarlier since pattern searching for is _ _ X where X is the period
```
this is the same comment as line 104. Is this one supposed to be X _ _ ?

kenneth-miura replied:
yes i copy pasted this is my b

faye-tan commented on tpp-app/src/services/cycle/CycleService.js:
```diff
@@ -0,0 +1,290 @@
+      calendar = await getCalendarByYear(date.getFullYear());
+    }
+
+    let mostRecentPeriodDay = getLastPeriodStart(date, calendar);
```
If we have a new user who has no periods logged yet, what will happen? I think the while loop in getLastPeriodStart() will be an infinite loop

faye-tan commented on tpp-app/src/services/cycle/CycleService.js:
```diff
@@ -0,0 +1,290 @@
+    var flowTwoDaysLater = (twoDaysLaterSymptoms.flow !== null && twoDaysLaterSymptoms.flow !== FLOW_LEVEL.NONE)
+
+
+    while(!(noFlowToday && noFlowTomorrow && flowTwoDaysLater)){
```
Correct me if I'm wrong but I think this while loop may not exit in some cases and be stuck in infinite loop. If we have a new user who has no period logged, this could be searching past dates forever for the _ _ X pattern, which does not exist. 

Also, I am not convinced subDays() is the best way of 'taking a step backwards' when iterating since you are always taking a step of one day backwards. A better way would be to take larger steps backwards by not checking dates where we know the pattern  _ _ X would not match. This strategy would be to identify dates where there are periods so we are guaranteed to match the "X" part of the pattern. 

What you have: 
- _ _ X is a pattern that walks backwards in time one day at a time 
- The strategy is not wrong, but there is a lot of unnecessary comparisons  
- If period dates in 2022 are [Jan 10, Jan 11, Jan 12, Jan 13], and searchFrom=Feb 28, you are making ~40 comparisons 

What I am suggesting: 
- Find list of period dates in current year using a helper. If year=2022 retrieve list of all period dates in 2022 ex: ```periodDates=[Jan 10, Jan 11, Jan 12, Jan 13]```
- searchFrom = Feb 28 
- _ _ X is a pattern that walks backwards in time by matching the X in the pattern to the list of ```periodDates```
  - check flowTwoDaysLater=Jan 13's flow, noFlowTomorrow=Jan 12's flow, noFlowToday=Jan 11's flow. It's not a match to _ _ X pattern.  Set the "X" in the pattern to the next date, Jan 12 to check Jan 10, Jan 11, Jan 12 if it matches _ _ X. 
  - And so on...until you get a match of _ _ X to Jan 8, Jan 9, Jan 10
  - This way instead of taking a step of one day backwards, we optimize by taking larger steps backwards to skip days where we know the pattern won't match giving us more of a chance of success in matching the desired pattern 
  - Instead of making ~40 comparisons, you only made 4 comparisons 

I think this would also solve the problem of the infinite while loop if there is no last period start since you can just go through ```periodDates```

We can chat/call about this, this is big infodump 💩

faye-tan replied:
Same logic can apply to ```getNextPeriodEnd()``` when searching for X _ _ pattern match

kenneth-miura replied:
so have a function (let's say `getPeriodDates`) that returns an array of Date objects that contains all the dates that there are periods in sequential order right? And then you only need to check for the pattern on these period dates returned from the function? 

Isn't that the same amount of comparisons at the end though? If we call the number of dates in a year _n_, then `getPeriodDates` has to do _n_ accesses to guarantee it has all the period days, and then do some more comparisons in `getLastPeriodStart` (O(n)). 

Also yeah you are totally right this does cause an infinite loop if the user has no period logged. I think swapping to ur idea would deal with the issue cleanly

kenneth-miura replied:
Actually I think it does mean markedly less computations for some functions since we can call `getPeriodDates` once and pass it in to the relevant functions in the cases where we call them multiple times (Basically just `GETCycleHistoryByYear`). 

The one-off functions like `GETPeriodDate` will be a bit slower if u don't give them a pre-computed list, but I think that's fine.

kenneth-miura replied:
Ok I refactored `GETPeriodDay` and `GETMostRecentPeriodStartDate` to use an array of periods for the year.

I also refactored `GETCycleHistoryByYear` to use an array of periods. I still use the step-by-step iteration approach for the very first and very last period of the year, so that it will be able to read periods like "Dec 27 2021" -> "Jan 2cnd 2022" as one interval in both `GETCycleHistoryByYear(2021)` and `GETCycleHistoryByYear(2022)`. I changed docstring of these step-by-step methods to be explicit that the `Date` to search from should be a period day, and wrote the code so they can't get called in a calendar that hasn't logged any periods.

kenneth-miura committed:
switched to periods array, partially working

swap worked for GETPeriodDays and GETMostRecentPeriodDate, GETCycleHistory currently doesn't include the first and last period of the year.

kenneth-miura committed:
working on calendar that spans years

kenneth-miura committed:
fixed double including first period if not spanning year

There was an error where, in a calendar in which the first period is entirely in the same year for GETCycleHistoryByYear, it would be included twice

kenneth-miura committed:
resolve todos, make variable names less bad

kenneth-miura committed:
removed testing + removed temp comments

kenneth-miura committed:
removing comments

faye-tan commented on tpp-app/src/services/cycle/CycleService.js:
```diff
@@ -0,0 +1,381 @@
+    let periods = await getPeriodsInYear(year, calendar);
+
+    if(periods.length == 0){
+      return intervals;
```
noice 👍

faye-tan commented on tpp-app/App.js:
```diff
@@ -102,6 +103,7 @@ export function MainPage() {
     // <Welcome></Welcome>
-     <MainPage></MainPage>
+    //  <MainPage></MainPage>
+    <DummyTest></DummyTest>
```
remove this before merge

faye-tan commented on tpp-app/src/services/cycle/CycleService.js:
```diff
@@ -0,0 +1,381 @@
+
+
+
+    let current = endOfYear;
```
unneeded now?

faye-tan commented on tpp-app/src/services/cycle/CycleService.js:
```diff
@@ -0,0 +1,381 @@
+ * @param {Object} calendar The object containing the symptoms for this year, last year, and next year. Optional.
+ * @return {Promise} resolves into a boolean, true if the date is a period start.
+ */
+async function isPeriodStart(date, calendar){
```
v nice, all this does is match the pattern

faye-tan commented on tpp-app/src/services/cycle/CycleService.js:
```diff
@@ -0,0 +1,381 @@
+    var flowTwoDaysEarlier = (twoDaysEarlierSymptoms.flow !== null && twoDaysEarlierSymptoms.flow !== FLOW_LEVEL.NONE)
+
+
+    while(!(noFlowToday && noFlowYesterday && flowTwoDaysEarlier)){
```
It seems like you are looking for _ _ X here. If you are looking for the end date of the final period in the next year shouldn't you be looking for X _ _? Or am I just confused 😅

faye-tan commented on tpp-app/src/services/cycle/CycleService.js:
```diff
@@ -0,0 +1,381 @@
+          isLastPeriodStart = false;
+        }
+        else{
+          if(isPeriodEnd){
```
nice! if X is first X seen after an interval, label it as the end of the period and don't mark any X as period end until after interval is completed.

good for cases such as _ _ X X X _ X _ X _ _

faye-tan reviewed (approved):
just some small comments, otherwise looks good!

helen-li reviewed (approved):
looking gooood ty kenneth !

kenneth-miura committed:
switched back to loading welcome page
